common/src/main/java/io/netty/util/concurrent/Future.java

```diff
@@ -49,12 +48,27 @@
     Throwable cause();
 
     /**
-     * Adds the specified listener to this future.  The
-     * specified listener is notified when this future is
-     * {@linkplain #isDone() done}.  If this future is already
-     * completed, the specified listener is notified immediately.
+     * Adds the specified listener to this future.
+     * The specified listener is notified when this future is {@linkplain #isDone() done}.
+     * If this future is already completed, the specified listener is notified immediately.
+     *
+     * @param listener The listener to be called when this future completes.
+     *                 The listener will be passed this future as an argument.
+     * @return this future object.
+     */
+    Future<V> addListener(FutureListener<? super V> listener);
+
+    /**
+     * Adds the specified listener to this future.
+     * The specified listener is notified when this future is {@linkplain #isDone() done}.
+     * If this future is already completed, the specified listener is notified immediately.
+     *
+     * @param context The context object that will be passed to the listener when this future completes.
+     * @param listener The listener to be called when this future completes.
+     *                 The listener will be passed the given context, and this future.
+     * @return this future object.
      */
-    Future<V> addListener(GenericFutureListener<? extends Future<? super V>> listener);
+    <C> Future<V> addListener(C context, FutureContextListener<? super C, ? super V> listener);
```

Comment:
I wonder if we really need this. Why not add something to FutureListener which allows to create a FutureListener that does the same.

Something like:

```
interface FutureListener {

    static <C, V> FutureListener newContextAwareListener(C context, BiConsumer<C, V> consumer);
}
```

Reply:
That approach will require more allocation. By passing the context in separately when adding the listener, the lambdas don't need to be capturing for the purpose of the context object at least.

Reply:
I see... I was just wondering if we should keep things "slim" in terms of API surface. Not strong on this tho.

Reply:
@chrisvest any thoughts about splitting the API?

Reply:
@NiteshKant Not a fan of splitting it. For one, I don't think the API is very complicated as is, and second, it found a lot of immediate usages when introduced here. I also briefly looked into #8523 today, and realised it can be used as a vehicle for future composition as well. Splitting the API as you've shown adds more moving parts, and adds a risk that some usage sites might once again care about the type of future, which was the reason we had `GenericFutureListener` and one of the main pain points this PR seeks to address. Not saying it's as bad with the `from()`, but it's a less clean break, IMHO.

Reply:
Fair enough, thanks for sharing your views!